### PR TITLE
[SPARK-50470][SQL] Block usage of collations for map keys

### PR DESCRIPTION
**Pull request by IAmAleksei**
### What changes were proposed in this pull request?

According to the issue description collation usage on keys of map may lead for unexpected effects (key duplication). This PR blocks this usage, but it is still available with Spark conf flag.
Old behavior is enabled by `spark.sql.collation.allowInMapKeys`.

### Why are the changes needed?

Because usage may create unclear situation. For example if we have `map('A' -> 1, 'a' -> 2)` then changing collation from UTF8_BINARY to UTF8_LCASE break key uniqueness.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by Alexvsalexvsalex** [comment]
> @Alexvsalexvsalex I'm fine with the changes mostly, but I just have a question whether it would be simpler to just disable altering the collations in the key than to completely forbid using them?

@stefankandic
Yeah, it was my first attempt. But the problem is in implicit schema changes (for example, append to existing table when a new column `map<string collate ..., ...>`  may be added from another table during schema merge).

**Review by stefankandic — APPROVED** [review]

**Comment by Alexvsalexvsalex** [comment]
@MaxGekk Could you please help with merge?

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -5188,6 +5188,11 @@
           "The SQL pipe operator syntax using |> does not support <clauses>."
         ]
       },
+      "COLLATIONS_IN_MAP_KEYS" : {
+        "message" : [
+          "Collated strings are not supported for keys of maps"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:5193]: Please, remove the duplicate `not supported`, see the full error message including parent's one:
```
The feature is not supported: Collated strings are not supported for keys of maps
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -1539,15 +1539,23 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
         alter.conf.resolver)
     }
 
+    def checkNoCollationsInMapKeys(colsToAdd: Seq[QualifiedColType]): Unit = {
+      if (!SQLConf.get.allowCollationsInMapKeys) {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:1543]: Should we use the config from the plan: `alter.conf.`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 3]
```diff
@@ -84,6 +85,9 @@ case class InsertIntoHadoopFsRelationCommand(
         outputColumnNames,
         sparkSession.sessionState.conf.caseSensitiveAnalysis)
     }
+    if (!SQLConf.get.allowCollationsInMapKeys) {
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala:88]: Take it from `conf`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 4]
```diff
@@ -338,6 +339,9 @@ case class PreprocessTableCreation(catalog: SessionCatalog) extends Rule[Logical
     SchemaUtils.checkSchemaColumnNameDuplication(
       schema,
       conf.caseSensitiveAnalysis)
+    if (!SQLConf.get.allowCollationsInMapKeys) {
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:342]:
```suggestion
    if (!conf.allowCollationsInMapKeys) {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 5]
```diff
@@ -241,6 +241,9 @@ case class AlterTableAddColumnsCommand(
     SchemaUtils.checkColumnNameDuplication(
       (colsWithProcessedDefaults ++ catalogTable.schema).map(_.name),
       conf.caseSensitiveAnalysis)
+    if (!SQLConf.get.allowCollationsInMapKeys) {
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:244]:
```suggestion
    if (!conf.allowCollationsInMapKeys) {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -767,6 +767,13 @@ object SQLConf {
       .checkValue(_ > 0, "The initial number of partitions must be positive.")
       .createOptional
 
+  lazy val ALLOW_COLLATIONS_IN_MAP_KEYS =
+    buildConf("spark.sql.collation.allowInMapKeys")
+      .doc("Allow usage of collated strings inside of map's keys")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:772]: any collated strings? I mean `str COLLATE UTF8_BINARY` is collated too. Should we say something like `Allow for non-UTF8_BINARY collated strings`

**Review by MaxGekk — COMMENTED** [review]
> Yes, but only in unreleased version with collation support.

Let's say `No` since the collation feature hasn't been released yet.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by Alexvsalexvsalex** [comment]
@MaxGekk thank you for your time. Tests are passed

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @Alexvsalexvsalex and @stefankandic for review.

**Pull request closed without merging**